romancal/skycell/match.py

```diff
@@ -0,0 +1,232 @@
+"""
+This module determines which sky cells overlap with the given image.
+
+Currently this assumes that the sky projected borders of all calibrated L2 images are straight.
+"""
+
+import logging
+from functools import cached_property
+
+import numpy as np
+import spherical_geometry.great_circle_arc as sga
+import spherical_geometry.polygon as sgp
+import spherical_geometry.vector as sgv
+from gwcs import WCS
+from numpy.typing import NDArray
+
+import romancal.skycell.skymap as sc
+from romancal.assign_wcs.utils import wcs_bbox_from_shape
+
+log = logging.getLogger(__name__)
+log.setLevel(logging.DEBUG)
+
+
+class ImageFootprint:
+    """abstraction of an image footprint"""
+
+    _radec_corners: NDArray[float]
+
+    def __init__(self, radec_corners: list[tuple[float, float]]):
+        """
+        Parameters
+        ----------
+        radec_corners: list[tuple[float, float]]
+            four corners of the image in right ascension and declination
+        """
+        radec_corners = np.array(radec_corners)
+        if radec_corners.shape[0] != 4:
+            raise ValueError(f"need 4 corners, not {radec_corners.shape[0]}")
+        self._radec_corners = radec_corners
+
+    @classmethod
+    def from_gwcs(
+        cls, iwcs: WCS, image_shape: tuple[int, int] | None = None
+    ) -> "ImageFootprint":
+        """create an image footprint from a GWCS object (and image shape, if no bounding box is present)"""
+
+        if iwcs.bounding_box is None:
+            if image_shape is None:
+                # Now must find size of corresponding image, with three possible sources of that information.
+                # Both bounding_box and pixel_shape are in x, y order contrary to numpy convention.
+                if hasattr(iwcs, "bounding_box") and iwcs.bounding_box is not None:
+                    bbintervals = iwcs.bounding_box.intervals
+                    # This compensates for the half pixel adjustment in the general code.
+                    image_shape = (
+                        bbintervals[1].upper + 0.5,
+                        bbintervals[0].upper + 0.5,
+                    )
+                elif hasattr(iwcs, "pixel_shape") and iwcs.pixel_shape is not None:
+                    image_shape = (iwcs.pixel_shape[1], iwcs.pixel_shape[0])
+                else:
+                    raise ValueError(
+                        "Cannot infer image footprint from GWCS object because "
+                        "`image_shape` was not specified and "
+                        "the GWCS object does not have `.bounding_box` nor `.pixel_shape`."
+                    )
+
+            iwcs.bounding_box = wcs_bbox_from_shape(image_shape)
```

Comment:
This roughly subtracts 0.5 from the shape.  Is that consistent with bbintervals adding 0.5?  I worry there is a fits / python convention difference in here.

Reply:
ah, I see now I did not fully understand @nden's point before. This shouldn't even be trying to derive the image shape from the bounding box here, since it's inside `iwcs.bounding_box is None`. I'll take out that `if` block